File: src/backend/TWAction.Domain/Settings/MainActionSettings.cs
namespace TWAction.Domain.Settings;

/// <summary>
/// Represents main action settings for a schedule.
/// </summary>
public sealed class MainActionSettings
{
    public Guid Id { get; set; }
    public Guid ScheduleId { get; set; }
    public DateTimeOffset MinDepartureTime { get; set; }
    public bool SkipNightSendings { get; set; }
    public uint MaxNobleDistance { get; set; }
    public MainActionOffSettings OffSettings { get; set; } = null!;
    public MainActionCatasSettings CatasSettings { get; set; } = null!;
    public MainActionFakeOffSettings FakeOffSettings { get; set; } = null!;
    public MainActionFakeDeffSettings FakeDeffSettings { get; set; } = null!;
    public MainActionNobleSettings NobleSettings { get; set; } = null!;
    public Dictionary<int, uint> PlayerNobleBudgets { get; set; } = new();
}

/// <summary>
/// Settings for off attacks.
/// </summary>
public sealed class MainActionOffSettings
{
    public uint MinOffUnits { get; set; }
    public uint MinDistanceFromFront { get; set; }
}

/// <summary>
/// Settings for catapult attacks.
/// </summary>
public sealed class MainActionCatasSettings
{
    public uint MinCatasNumber { get; set; }
    public uint MinDistanceFromFront { get; set; }
    public uint MaxOffUnits { get; set; }
}

/// <summary>
/// Settings for fake off attacks.
/// </summary>
public sealed class MainActionFakeOffSettings
{
    public uint MinOffUnits { get; set; }
    public uint MinDistanceFromFront { get; set; }
}

/// <summary>
/// Settings for fake deff attacks.
/// </summary>
public sealed class MainActionFakeDeffSettings
{
    public uint MaxOffUnits { get; set; }
    public uint MinDistanceFromFront { get; set; }
}

/// <summary>
/// Settings for noble attacks.
/// </summary>
public sealed class MainActionNobleSettings
{
    public uint MinDistanceFromFront { get; set; }
    public uint MinOffUnitsForOffNoble { get; set; }
    public uint MinOffUnitsForFakeOffNoble { get; set; }
    public uint MaxOffUnitsForDefNoble { get; set; }
    public uint MinDeffUnitsForDefNoble { get; set; }
}

File: src/frontend/TWActionFrontend/src/types/mainActionSettings.ts
export interface MainActionOffSettings {
  minOffUnits: number;
  minDistanceFromFront: number;
}

export interface MainActionCatasSettings {
  minCatasNumber: number;
  minDistanceFromFront: number;
  maxOffUnits: number;
}

export interface MainActionFakeOffSettings {
  minOffUnits: number;
  minDistanceFromFront: number;
}

export interface MainActionFakeDeffSettings {
  maxOffUnits: number;
  minDistanceFromFront: number;
}

export interface MainActionNobleSettings {
  minDistanceFromFront: number;
  minOffUnitsForOffNoble: number;
  minOffUnitsForFakeOffNoble: number;
  maxOffUnitsForDefNoble: number;
  minDeffUnitsForDefNoble: number;
}

export interface MainActionSettings {
  id: string;
  scheduleId: string;
  minDepartureTime: string;
  skipNightSendings: boolean;
  maxNobleDistance: number;
  offSettings: MainActionOffSettings;
  catasSettings: MainActionCatasSettings;
  fakeOffSettings: MainActionFakeOffSettings;
  fakeDeffSettings: MainActionFakeDeffSettings;
  nobleSettings: MainActionNobleSettings;
  playerNobleBudgets: Record<number, number>;
}

export interface SaveMainActionSettingsRequest {
  minDepartureTime: string;
  skipNightSendings: boolean;
  maxNobleDistance: number;
  offSettings: MainActionOffSettings;
  catasSettings: MainActionCatasSettings;
  fakeOffSettings: MainActionFakeOffSettings;
  fakeDeffSettings: MainActionFakeDeffSettings;
  nobleSettings: MainActionNobleSettings;
  playerNobleBudgets: Record<number, number>;
}

export const getDefaultMainActionSettings = (): Omit<
  SaveMainActionSettingsRequest,
  "minDepartureTime"
> & {
  minDepartureTime: Date;
} => {
  const now = new Date();

  // Domyślna data to następny dzień 8 rano
  const minDepartureTime = new Date(now);
  minDepartureTime.setDate(minDepartureTime.getDate() + 1);
  minDepartureTime.setHours(8, 0, 0, 0);

  return {
    minDepartureTime,
    skipNightSendings: true,
    maxNobleDistance: 49,
    offSettings: {
      minOffUnits: 18000,
      minDistanceFromFront: 5,
    },
    catasSettings: {
      minCatasNumber: 50,
      minDistanceFromFront: 5,
      maxOffUnits: 25000,
    },
    fakeOffSettings: {
      minOffUnits: 10000,
      minDistanceFromFront: 5,
    },
    fakeDeffSettings: {
      maxOffUnits: 10000,
      minDistanceFromFront: 5,
    },
    nobleSettings: {
      minDistanceFromFront: 5,
      minOffUnitsForOffNoble: 10000,
      minOffUnitsForFakeOffNoble: 7000,
      maxOffUnitsForDefNoble: 10000,
      minDeffUnitsForDefNoble: 10000,
    },
    playerNobleBudgets: {},
  };
};
